### PR TITLE
Implements trigger-action bindings

### PR DESCRIPTION
**Pull request by tarbayev**
**Trigger-action** bindings allow to set up binding which triggers action on target object in response to observable value change.
Optimizes binding querying. Removes `-[NSObject mfb_bindingsForKeyPath:]` query method considered confusing and redundant.
Binding query methods are only available in Debug configuration.
Improves binding assertion tests by using additional target for testing macros with foundation assertions disabled.

### CODE REVIEW

**Comment by nikolaykasyanov** [comment]
On it, it'd be great to update CocoaPods separately in the future (too much diff noise).

**Comment by tarbayev** [comment]
@nikolaykasyanov my fault, had some issues with linting, tried update first and realized the real issue later.

File: Sources/MFBActionBinding.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+//
+//
+
+#import <Foundation/Foundation.h>
+
+#import "NSObject+MFBBindingAssertion.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MFBActionBinding : NSObject
```
Comment (nikolaykasyanov) [inline Sources/MFBActionBinding.h:15]: I wonder if we could abuse UIControl in order to be able just drug "Value Changed" event to an IBAction.

File: Sources/MFBActionBinding.m [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+static NSString *const TriggeringBindingsGroup = @"TriggeringBindings";
+static NSString *const ActionBindingsGroup = @"ActionBindings";
+
+@interface MFBActionBinding () <MFBUnbindable>
+@property (nonatomic) SEL actionSelector;
+@end
+
+@implementation MFBActionBinding {
+    BOOL _isBound;
```
Comment (nikolaykasyanov) [inline Sources/MFBActionBinding.m:25]: These two flags could be replaced with one enum of `NotBound`, `Bound`, `BoundAndUpdating`. This rules out invalid states like "not bound, but updating" statically.

File: Sources/MFBBinding.m [hunk 3]
```diff
@@ -300,48 +34,26 @@ - (void)awakeFromNib
...
-
-    _firstKeyPath = [firstKeyPath copy];
-
-    [self registerIfValid];
-}
-
 - (void)setSecondObject:(id)secondObject
 {
+    NSCParameterAssert(_secondKeyPath != nil);
+    NSCParameterAssert(!_secondObject);
```
Comment (nikolaykasyanov) [inline Sources/MFBBinding.m]: Explicit comparison, please.

File: Sources/NSObject+MFBBindingAssertion.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+#import <Foundation/Foundation.h>
+
+@interface NSObject (MFBBindingAssertion)
+
+/**
+ Setting this property to @p YES will disable assertions made by @a MFBAssertGetterBinding and @a MFBAssertSetterBinding
+ macros for the receiver.
+ */
+@property (nonatomic) BOOL bindingAssertionDisabled;
```
Comment (nikolaykasyanov) [inline Sources/NSObject+MFBBindingAssertion.h:17]: Prefix, please.

File: Sources/Private/MFBBindingInfo.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+@end
+
+@interface MFBBindingInfo : NSObject
+
++ (instancetype)bindingInfoForObject:(id)obj;
+
+- (void)addBinding:(id<MFBUnbindable>)binding forKey:(NSString *)key group:(NSString *)group;
+- (void)removeBinding:(id<MFBUnbindable>)binding forKey:(NSString *)key group:(NSString *)group;
+
+- (NSArray *)bindingsForKey:(NSString *)key inGroup:(NSString *)group;
```
Comment (nikolaykasyanov) [inline Sources/Private/MFBBindingInfo.h:24]: Does generic type make sense here?

**Comment by tarbayev** [comment]
@nikolaykasyanov Also we could update CocoaPods and rebase.

File: Tests/MFBActionBindingTests.m [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+
+#pragma mark - Value Binding Test Methods
+
+- (void)test_observablePropertyChange_TriggersActionOnTarget
+{
+    [_configuration setUpBinding];
+
+    for (id v in @[ [NSObject new], [NSNull null] ]) {
+
+        const id Value = v != [NSNull null] ? v : nil;
```
Comment (nikolaykasyanov) [inline Tests/MFBActionBindingTests.m]: I'd go for brackets whenever ternary operator condition is not just variable.

File: .travis.yml [hunk 7]
```diff
@@ -14,10 +14,14 @@ branches:
...
-    - DESTINATION="platform=iOS Simulator,OS=8.1,name=iPhone 5" #32-bit
-    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 5" #32-bit
-    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 6s" #64-bit
+    - DESTINATION="platform=iOS Simulator,OS=8.1,name=iPhone 5" SCHEME="MFBBindingTests_Debug" #32-bit
+    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 5" SCHEME="MFBBindingTests_Debug" #32-bit
+    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 6s" SCHEME="MFBBindingTests_Debug" #64-bit
+
+    - DESTINATION="platform=iOS Simulator,OS=8.1,name=iPhone 5" SCHEME="MFBBindingTests_Release" #32-bit
+    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 5" SCHEME="MFBBindingTests_Release" #32-bit
+    - DESTINATION="platform=iOS Simulator,OS=9.3,name=iPhone 6s" SCHEME="MFBBindingTests_Release" #64-bit
```
Comment (nikolaykasyanov) [inline .travis.yml]: What about adding 10.3 (and probably dropping 8.1)?

**Review by nikolaykasyanov — CHANGES_REQUESTED** [review]

File: Sources/MFBActionBinding.h (hunk 1 above)
Comment (tarbayev) [reply Sources/MFBActionBinding.h:15]: I find it a bit too hacky. UIControl has a long inheritance chain not worth to deal with. Also we cannot limit UIControlEvents list available from IB.

File: Sources/NSObject+MFBBindingAssertion.h (hunk 4 above)
Comment (tarbayev) [reply Sources/NSObject+MFBBindingAssertion.h:17]: Didn't want to introduce one more breaking change. Better to make it in separate PR.

Comment (nikolaykasyanov) [reply Sources/NSObject+MFBBindingAssertion.h:17]: I see, fair.

File: Sources/MFBActionBinding.h (hunk 1 above)
Comment (nikolaykasyanov) [reply Sources/MFBActionBinding.h:15]: Right.

File: .travis.yml (hunk 7 above)
Comment (tarbayev) [reply .travis.yml]: I'll add 10.3, however as far as lib supports iOS 8, I wouldn't drop anything.

File: Sources/MFBActionBinding.m (hunk 2 above)
Comment (tarbayev) [reply Sources/MFBActionBinding.m:25]: But we would need to check for `Bound` before transitioning to `BoundAndUpdating` anyway, same we could do with boolean flags.

File: Sources/MFBBinding.m (hunk 3 above)
Comment (tarbayev) [reply Sources/MFBBinding.m]: Why? Apple's convention states for `!obj` for checking for `nil`.

Comment (nikolaykasyanov) [reply Sources/MFBBinding.m]: Because it's an assert?

Comment (nikolaykasyanov) [reply Sources/MFBBinding.m]: and we agreed to always use explicit comparisons in asserts.

File: Sources/MFBActionBinding.m (hunk 2 above)
Comment (nikolaykasyanov) [reply Sources/MFBActionBinding.m:25]: Okay.

File: Sources/Private/MFBBindingInfo.h (hunk 5 above)
Comment (tarbayev) [reply Sources/Private/MFBBindingInfo.h:24]: Bindings being added or removed have to be unbindable (`MFBUnbindable`). Bindings returned for key and group pair are just objects without any specific requirement as clients should not unbind them.

File: Tests/MFBActionBindingTests.m (hunk 6 above)
Comment (tarbayev) [reply Tests/MFBActionBindingTests.m]: Decided to go with generator function.

File: Sources/Private/MFBBindingInfo.h (hunk 5 above)
Comment (nikolaykasyanov) [reply Sources/Private/MFBBindingInfo.h:24]: makes sense

**Review by nikolaykasyanov — APPROVED** [review]

File: Sources/Private/MFBBindingInfo.h (hunk 5 above)
Comment (nikolaykasyanov) [reply Sources/Private/MFBBindingInfo.h:24]: Actually, as far as I can see, the only thing we do with the result of this method is invoking `count` and comparing it to `0`. Therefore, could its return type be changed to just `BOOL`?

Comment (tarbayev) [reply Sources/Private/MFBBindingInfo.h:24]: In asserts we only use count, however for debug purposes one might use the complete list of bindings.

Comment (nikolaykasyanov) [reply Sources/Private/MFBBindingInfo.h:24]: 👍🏻
btw, if we want to make objects in array completely opaque, wouldn't it be better to use array of `NSObject`? One can call anything on `id`.

Comment (nikolaykasyanov) [reply Sources/Private/MFBBindingInfo.h:24]: on the other hand, we can't prevent shooting oneself in the foot completely anyway ¯\_(ツ)_/¯

Comment (tarbayev) [reply Sources/Private/MFBBindingInfo.h:24]: Again, it's for debug purposes only, thus one can call whatever he wants without the need to cast it to `id` every time.

**Pull request merged**